### DISCUSSION TITLE
Distance not filled

### DISCUSSION THREAD

**Discussion by Werzi2001** [magicbug/Cloudlog#2202] [General]
Hello everybody,

I just noticed that there actually is a COL_DISTANCE in the database which is also available in the advanced search. Anyhow that column is not filled (always NULL) and the distance shown in the UI (e.g. in the QSO details) seems to be calculated on-the-fly.

I would find it really interesting to have that column filled correctly (especially to search by it using the advanced search). Is there any specific reason why the column exists but is not filled? I'm not really sure where it should be filled (maybe every time a QSO is created/updated?) but if you give me a hint I would try to add that feature.

Best Regards
Werzi2001

**Comment by magicbug** [comment]
Legacy field that hasn't been used since the early days of development, its far easier todo it on the fly

**Reply by Werzi2001** [reply]
Yes, as long as you just wont to display it for some/a few QSOs. If you want to do something like "give me all QSOs that were confirmed by LotW and have a distance >2500km" it gets difficult to calculate on-the-fly.

But I understand that it's legacy and there is no plan to fill it. Maybe I'll just create a script then that sets it for all QSOs (like the continent update).

**Reply by phl0** [reply]
Well there is code that calculates the needed value upon entering the grid:

![Screenshot from 2023-06-14 22-29-21](https://github.com/magicbug/Cloudlog/assets/7112907/241fb7f2-1855-4977-8909-30da22887f12)

See function `get_bearing` in application/libraries/Qra.php.

**Comment by phl0** [comment]
Guys I made a small PoC that uses the calculated distance and stores it into the QSO table. This is just PoC state as of now. I.e. only storing upon creating the QSO. No UI for update/editing of this value (yet). Just looking for feedback here. See:

https://github.com/phl0/Cloudlog/tree/storeDistance

**Comment by magicbug** [comment]
Nice idea but would need logic to handle if the grid changes at any point within Cloudlog think this would only be qso edit and perhaps lotw

**Reply by phl0** [reply]
Logic added to handle grid changes. QSO edit still to go. 
Cannot see a place where this would affect LoTW though.

**Reply by magicbug** [reply]
If i remember correct if Grid is missing and its available via LoTW cloudlog will store it hence why I thought we'd need to run distance calc to store on that

**Reply by phl0** [reply]
Ah good point. Will check that (also).

**Reply by magicbug** [reply]
I could be wrong that it does it but I think it does lol

**Reply by phl0** [reply]
Nope. You were right. So added some lines of code (and another SQL select) to calc distance upon LoTW match/update. Feedback welcome.

**Reply by magicbug** [reply]
Without trying the branch the code seems ok

**Reply by Werzi2001** [reply]
In the meantime I wrote an external script to fill the distance properly (I can just reassure that it is very helpful in combination with the advanced search). Anyhow I'm very happy that @phl0 is working on it. Thanks a lot :)
I guess a very annoying case would be if the grid of the station profile changes. As in that case all distances within the profile might have to be updated. Although I'm not sure if that is really a use case. Probably only if the initial grid was incorrect.
For me the distance is actually a very interesting information. I would even like to add it to the logbook main list as part of the configurable columns. Anyhow I can add that feature myself as soon as you finished filling the distance :)

**Reply by phl0** [reply]
I did update the branch a little so that upon editing a QSO and changing the grid the distance is also updated. Plus it shows the distance also below the grid field in edit QSO.
Next up: A global update function to (re-)calculate all distances in database where distance is zero or not set yet. I might add an option to force recalc of all distances to cover the use case where someone changes the grid of a station location.

**Reply by magicbug** [reply]
Last bit sounds like a great idea Flo!

**Comment by phl0** [comment]
My code is now merged into dev branch. @Werzi2001 pls test and report.

**Reply by Werzi2001** [reply]
@phl0 Thanks a lot. Worked fine in my logbook. Just one thing I noticed: I had some QSOs which a distance of 0 instead of null before the change already. These were not updated by "Update distance data". Not sure where these came from but I don't think I added these somehow. Maybe an older version set distance to 0 instead of null? Might make sense to update these as well.

Furthermore I created the pull request #2216 to add the possibility to select the distance as a column in the logbook etc.

**Reply by Werzi2001** [reply]
Just noticed one more thing @phl0 :
Removing the gridsquare from a QSO (or entering an incorrect gridsquare) leads to a PHP error (and QSO cannot be saved. This happens for me sometimes if a gridsquare is rertieved from some source (e.g. QRZ) but I know that it is not right (e.g. because it was a portable QSO).

**Reply by phl0** [reply]
I will look into that. Tnx for notice.

**Reply by phl0** [reply]
Could not reproduce that one. JS is hindering me to save the QSO with invalid grid e.g. JO1. Can you tell me how to repoduce please?

**Reply by Werzi2001** [reply]
Yes I can also not save the QSO but I get a php error message in the edit QSO screen:
![screenshot](https://github.com/magicbug/Cloudlog/assets/12096649/b0db70d1-d8d5-4aa0-b0fe-4f03795eb85e)

**Reply by Werzi2001** [reply]
I have to add that I can't save because of a php error:
A Database Error Occurred
Error Number: 1366
Incorrect double value: '

Not because of a Javascript check/error. I think it should be possible to save empty gridsquare (and am pretty sure it worked before).

**Reply by phl0** [reply]
I am too dumb to reproduce this. I log a QSO with empty gridsquare (works). Afterwards I edit the QSO, change nothing and save again. No errors seen. Where am I going wrong pls?

**Reply by Werzi2001** [reply]
What I do is: Log an QSO with a gridsquare, edit the QSO and remove the gridsquare. As soon as I remove the value and leave the field the error appears. Also I can't save the change because of the php error on save.
Seems to be only a notice level error though. Maybe you have different php settings ignoring notices?

**Reply by magicbug** [reply]
Interesting tried here with all errors turned on (using the dev branch) and when I edit a qso and remove the square although I get 0km which should really just be not shown I don't get the PHP error

Can you double check your using the latest copy of dev branch from magicbug/cloudlog

**Reply by phl0** [reply]
Good point. I will add an if to suppress the message if result is 0.

**Reply by Werzi2001** [reply]
I just tried it again with the current `dev` branch and I still get the message. As far as I can see a call to `/index.php/logbook/searchdistance//10` is made. As you can see the gridsquare is empty (as I removed the value from the field). That leads to the PHP notice error.

Edit:
Checked the code and the calls seems to be done from `/assets/js/sections/common.js#L143`. There is no check if the value is non-empty and therefore the call is done without a grid square. On the other hand in `/assets/js/sections/qso.js#L694` is a check that the value is non-empty. I guess that should be added in common.js too?

**Reply by AndreasK79** [reply]
I see the call too, but no error. In fact, it returns 6804.5... go figure. The check can be added in common.js too then, to stop the call.

**Reply by AndreasK79** [reply]
Pushed a change to common.js.

**Comment by magicbug** [comment]
Seems OK just ran it

**Comment by phl0** [comment]
> Seems OK just ran it

Tnx feedback. The initial run to update all QSOs with missing data may run for a while on first run :-)

**Reply by magicbug** [reply]
14000 qsos went though pretty quick

**Comment by phl0** [comment]
> Just one thing I noticed: I had some QSOs which a distance of 0 instead of null before the change already. These were not updated by "Update distance data". Not sure where these came from but I don't think I added these somehow. Maybe an older version set distance to 0 instead of null? Might make sense to update these as well.

Hhmmmm difficult question. You could log a QSO with a station in the same square. Distance is then 0 and stored as such. These QSOs would be updated with 0s over and over again ... This is why I decided to only select QSOs with no distance at all.

> Furthermore I created the pull request #2216 to add the possibility to select the distance as a column in the logbook etc.

Thats great.
